### ISSUE TITLE
VB Language constraint "On Error Resume Next" vs lambda not reported in IDE

### ISSUE THREAD

**Issue by gafter** [dotnet/roslyn#3675]
The following program is an error according to the VB language specification (a method contains both "On Error Goto Label" statement and a lambda - see VB language spec 10.10.2.2). However the Roslyn VB compiler does not report this semantic error (from `Compilation.GetDiagnostics`), and the IDE fails to report the issue in the error log, even after the build fails. The Roslyn compiler does, however, report the error when attempting to build/emit.

As a rule, all *language semantic errors* should be reported through `Compilation.GetDiagnostics`.

```vb
Module Module1

    Sub Main()
        Dim item = Nothing
        If item Is Nothing Then
            On Error GoTo Foo
        End If
        If item IsNot Nothing Then
            Dim query = Function() item
        End If
Foo:
    End Sub

End Module
```

**Comment by gafter** [comment]
My temptation is to close this as “will not fix”. The “On Error GoTo” statement is rarely used today, so there is little value in spending effort perfecting the diagnostic experience for its misuse.

However, there is a corresponding issue with the much-more-frequently used `GoTo` statement:

```vb
Class D
    Public Shared Sub Main()
        GoTo Foo
        If True Then
            Dim j = 0
Foo:
            Dim query = Function() j
        End If
    End Sub
End Class
```

The required error here

```none
error BC36597: 'Goto Foo' is not valid because 'Foo' is inside a scope that defines a variable that is used in a lambda or query expression.
```
Is only produced during emit.

**Comment by gafter** [comment]
@heejaechang This is related to #1400.
/cc @AlekseyTs 

We cannot reasonably make the compiler produce these during `compilation.GetDiagnostics()`. On the other hand, we don't believe it will be possible for the compiler to cleanly divide the diagnostics it produces into the categories of "*those that `GetDiagnostics` produces*" and "*those that are produced by a build but not by `GetDiagnostics`*". There are some diagnostics (like "Missing type") that may be produced from `GetDiagnsotics()`, but can also be produced during `Emit` due to missing assembly references for helper types.

Probably the best we could do is to tag the individual diagnostics with the "phase" they came from - Parse, Binding, or Emit - so that the IDE can treat the individual diagnostics appropriately. There is no API currently in `Diagnostic` that appears to be appropriate for this. What do you recommend?
